src/allmydata/bsd/inotify.py

```diff
@@ -0,0 +1,163 @@
+[humanReadableMask, \
+    IN_WATCH_MASK, IN_ACCESS, IN_MODIFY, IN_ATTRIB, IN_CLOSE_NOWRITE, IN_CLOSE_WRITE, \
+    IN_OPEN, IN_MOVED_FROM, IN_MOVED_TO, IN_CREATE, IN_DELETE, IN_DELETE_SELF, \
+    IN_MOVE_SELF, IN_UNMOUNT, IN_Q_OVERFLOW, IN_IGNORED, IN_ONLYDIR, IN_DONT_FOLLOW, \
+    IN_MASK_ADD, IN_ISDIR, IN_ONESHOT, IN_CLOSE, IN_MOVED, IN_CHANGED]
+
+
+TRUE  = 0
+FALSE = 1
+
+NOT_STARTED = "NOT_STARTED"
+STARTED     = "STARTED"
+STOPPING    = "STOPPING"
+STOPPED     = "STOPPED"
+
+class INotifyEventHandler(FileSystemEventHandler):
+
+    def __init__(self, path, mask, callbacks, pending_delay):
+        print "init INotifyEventHandler"
+        FileSystemEventHandler.__init__(self)
+        self._path = path
+        self._mask = mask
+        self._callbacks = callbacks
+        self._pending_delay = pending_delay
+        self._pending = set()
+
+    def is_masked(self, event):
+        if isinstance(event, FileSystemMovedEvent) and self._mask & (IN_MOVED_TO | IN_MOVED_FROM):
+            return True
+        if (isinstance(event, FileModifiedEvent) or isinstance(event, DirModifiedEvent)) and self._mask & (IN_CLOSE_WRITE | IN_CHANGED):
+            return True
+        if (isinstance(event, FileCreatedEvent) or isinstance(event, DirCreatedEvent)) and self._mask & IN_CREATE:
+            return True
+        if (isinstance(event, FileDeletedEvent) or isinstance(event, DirDeletedEvent)) and self._mask & IN_DELETE:
+            return True
+        return False
+
+    def get_event_mask(self, event):
+        if event.is_directory:
+            mask = IN_ISDIR
+        else:
+            mask = 0 # XXX
+        if isinstance(event, FileModifiedEvent) or isinstance(event, DirModifiedEvent):
+            mask = IN_CHANGED
+        if isinstance(event, FileDeletedEvent) or isinstance(event, DirDeletedEvent):
+            mask = IN_DELETE
+        if isinstance(event, FileCreatedEvent) or isinstance(event, DirCreatedEvent):
+            mask = IN_CREATE
+        return mask
+
+    def process(self, event):
+        event_filepath_u = event.src_path.decode('utf-8')
+        if event_filepath_u == unicode_from_filepath(self._path):
+            # ignore events for parent directory
+            return
+        #if not self.is_masked(event):
+        #    return
+        #try:
+        #    event_path = self._path.preauthChild(event.src_path)
+        #except InsecurePath, e:
+        #    print "failed: %r" % (e,)
+        #    return
+
+        def _maybe_notify(path):
+            if path in self._pending:
+                return
+            self._pending.add(path)
+            def _do_callbacks():
+                print "DO CALLBACKS"
+                self._pending.remove(path)
+                #event_mask = self.get_event_mask(event)
+                event_mask = IN_CHANGED
+                for cb in self._callbacks:
+                    try:
+                        cb(None, FilePath(path), event_mask)
```

Comment:
Here we're relying for security on the path that FSEvents gives us to actually be relative and not escape to above the magic folder directory. On Windows we use ``preauthChild`` to avoid that reliance.